File: packages/java/src/main/java/io/constellationnetwork/metagraph/sdk/Verify.java
package io.constellationnetwork.metagraph.sdk;

import org.bouncycastle.crypto.params.ECDomainParameters;
import org.bouncycastle.crypto.params.ECPublicKeyParameters;
import org.bouncycastle.crypto.signers.ECDSASigner;
import org.bouncycastle.math.ec.ECPoint;

import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;

/**
 * Signature verification using ECDSA secp256k1.
 */
public final class Verify {

    // secp256k1 curve order for signature normalization
    private static final BigInteger SECP256K1_N = new BigInteger(
        "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141", 16
    );
    private static final BigInteger SECP256K1_HALF_N = SECP256K1_N.shiftRight(1);

    private Verify() {
        // Utility class
    }

    /**
     * Verify a signed object.
     *
     * @param signed Signed object with value and proofs
     * @param isDataUpdate Whether the value was signed as a DataUpdate
     * @return VerificationResult with valid/invalid proof lists
     */
    public static <T> Types.VerificationResult verify(Types.Signed<T> signed, boolean isDataUpdate) {
        // Compute the hash that should have been signed
        byte[] bytes = Binary.toBytes(signed.getValue(), isDataUpdate);
        Types.Hash hash = Hash.hashBytes(bytes);

        List<Types.SignatureProof> validProofs = new ArrayList<>();
        List<Types.SignatureProof> invalidProofs = new ArrayList<>();

        for (Types.SignatureProof proof : signed.getProofs()) {
            try {
                boolean isValid = verifyHash(hash.getValue(), proof.getSignature(), proof.getId());
                if (isValid) {
                    validProofs.add(proof);
                } else {
                    invalidProofs.add(proof);
                }
            } catch (Exception e) {
                invalidProofs.add(proof);
            }
        }

        boolean isValid = invalidProofs.isEmpty() && !validProofs.isEmpty();
        return new Types.VerificationResult(isValid, validProofs, invalidProofs);
    }

    /**
     * Verify a signature against a SHA-256 hash.
     *
     * <p>Protocol:
     * <ol>
     *   <li>Treat hash hex as UTF-8 bytes (NOT hex decode)</li>
     *   <li>SHA-512 hash</li>
     *   <li>Truncate to 32 bytes</li>
     *   <li>Verify ECDSA signature</li>
     * </ol>
     *
     * @param hashHex SHA-256 hash as 64-character hex string
     * @param signature DER-encoded signature in hex format
     * @param publicKeyId Public key in hex (with or without 04 prefix)
     * @return true if signature is valid
     */
    public static boolean verifyHash(String hashHex, String signature, String publicKeyId) {
        try {
            // Normalize and parse public key
            String fullPublicKey = Wallet.normalizePublicKey(publicKeyId);
            byte[] publicKeyBytes = Wallet.hexToBytes(fullPublicKey);

            ECDomainParameters ecParams = Wallet.getEcParams();
            ECPoint publicPoint = ecParams.getCurve().decodePoint(publicKeyBytes);
            ECPublicKeyParameters pubKeyParams = new ECPublicKeyParameters(publicPoint, ecParams);

            // Parse signature
            byte[] signatureBytes = Wallet.hexToBytes(signature);
            BigInteger[] rs = decodeDER(signatureBytes);
            BigInteger r = rs[0];
            BigInteger s = rs[1];

            // Normalize S to low-S form for verification compatibility
            if (s.compareTo(SECP256K1_HALF_N) > 0) {
                s = SECP256K1_N.subtract(s);
            }

            // Compute signing digest
            byte[] digest = Hash.computeDigestFromHash(hashHex);

            // Verify signature
            ECDSASigner verifier = new ECDSASigner();
            verifier.init(false, pubKeyParams);
            return verifier.verifySignature(digest, r, s);

        } catch (Exception e) {
            return false;
        }
    }

    /**
     * Verify a single signature proof against data.
     *
     * @param data The original data that was signed
     * @param proof The signature proof to verify
     * @param isDataUpdate Whether data was signed as DataUpdate
     * @return true if signature is valid
     */
    public static boolean verifySignature(Object data, Types.SignatureProof proof, boolean isDataUpdate) {
        byte[] bytes = Binary.toBytes(data, isDataUpdate);
        Types.Hash hash = Hash.hashBytes(bytes);
        return verifyHash(hash.getValue(), proof.getSignature(), proof.getId());
    }

    /**
     * Decode DER-encoded signature to (r, s) components.
     */
    private static BigInteger[] decodeDER(byte[] der) {
        if (der[0] != 0x30) {
            throw new Types.SdkException("Invalid DER signature format");
        }

        int offset = 2; // Skip 0x30 and total length

        // Parse R
        if (der[offset] != 0x02) {
            throw new Types.SdkException("Invalid DER signature: expected 0x02 for R");
        }
        int rLen = der[offset + 1] & 0xFF;
        byte[] rBytes = new byte[rLen];
        System.arraycopy(der, offset + 2, rBytes, 0, rLen);
        BigInteger r = new BigInteger(1, rBytes);
        offset += 2 + rLen;

        // Parse S
        if (der[offset] != 0x02) {
            throw new Types.SdkException("Invalid DER signature: expected 0x02 for S");
        }
        int sLen = der[offset + 1] & 0xFF;
        byte[] sBytes = new byte[sLen];
        System.arraycopy(der, offset + 2, sBytes, 0, sLen);
        BigInteger s = new BigInteger(1, sBytes);

        return new BigInteger[]{r, s};
    }
}

File: packages/typescript/src/verify.ts
/**
 * Signature Verification
 *
 * Verify ECDSA signatures using secp256k1 curve via @noble/curves.
 */

import { sha256 } from '@noble/hashes/sha256';
import { bytesToHex } from '@noble/curves/abstract/utils';
import { Signed, SignatureProof, VerificationResult } from './types';
import { toBytes } from './binary';
import { constellationDigest, ecdsaVerify } from './crypto';

/**
 * Verify a signed object
 *
 * When the signed object has a `mode` field (set by createSignedObject/addSignature/batchSign),
 * verification automatically uses the correct mode. The `isDataUpdate` parameter is only
 * needed for objects without a `mode` field (e.g., deserialized from older wire format).
 *
 * @param signed - Signed object with value and proofs
 * @param isDataUpdate - Whether the value was signed as a DataUpdate (ignored if signed.mode is set)
 * @returns VerificationResult with valid/invalid proof lists
 *
 * @example
 * ```typescript
 * // Mode is auto-detected from signed object
 * const result = verify(signedObject);
 * if (result.isValid) {
 *   console.log('All signatures valid');
 * }
 * ```
 */
export function verify<T>(
  signed: Signed<T>,
  isDataUpdate: boolean = false
): VerificationResult {
  // Prefer mode from signed object; fall back to isDataUpdate parameter
  const useDataUpdate = signed.mode
    ? signed.mode === 'dataUpdate'
    : isDataUpdate;
  const bytes = toBytes(signed.value, useDataUpdate);
  const hashHex = bytesToHex(sha256(bytes));

  const validProofs: SignatureProof[] = [];
  const invalidProofs: SignatureProof[] = [];

  for (const proof of signed.proofs) {
    try {
      const isValid = verifyHash(hashHex, proof.signature, proof.id);
      if (isValid) {
        validProofs.push(proof);
      } else {
        invalidProofs.push(proof);
      }
    } catch {
      invalidProofs.push(proof);
    }
  }

  return {
    isValid: invalidProofs.length === 0 && validProofs.length > 0,
    validProofs,
    invalidProofs,
  };
}

/**
 * Verify a signature against a SHA-256 hash
 *
 * Protocol:
 * 1. Treat hash hex as UTF-8 bytes (NOT hex decode)
 * 2. SHA-512 hash
 * 3. Truncate to 32 bytes
 * 4. Verify ECDSA signature (with low-S normalization)
 *
 * @param hashHex - SHA-256 hash as 64-character hex string
 * @param signature - DER-encoded signature in hex format
 * @param publicKeyId - Public key in hex (with or without 04 prefix)
 * @returns true if signature is valid
 */
export function verifyHash(
  hashHex: string,
  signature: string,
  publicKeyId: string
): boolean {
  try {
    const fullPublicKey = normalizePublicKey(publicKeyId);
    const digest = constellationDigest(hashHex);
    return ecdsaVerify(digest, signature, fullPublicKey);
  } catch {
    return false;
  }
}

/**
 * Verify a single signature proof against data
 *
 * @param data - The original data that was signed
 * @param proof - The signature proof to verify
 * @param isDataUpdate - Whether data was signed as DataUpdate
 * @returns true if signature is valid
 */
export function verifySignature<T>(
  data: T,
  proof: SignatureProof,
  isDataUpdate: boolean = false
): boolean {
  const bytes = toBytes(data, isDataUpdate);
  const hashHex = bytesToHex(sha256(bytes));
  return verifyHash(hashHex, proof.signature, proof.id);
}

/**
 * Normalize public key to full format (with 04 prefix)
 */
function normalizePublicKey(publicKey: string): string {
  if (publicKey.length === 128) {
    return '04' + publicKey;
  }
  if (publicKey.length === 130 && publicKey.startsWith('04')) {
    return publicKey;
  }
  return publicKey;
}
